_health-care/_js/_components/date_input.jsx

```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import _ from 'lodash';
+
+class DateInput extends React.Component {
+  constructor() {
+    super();
+    this.state = {month: '', date: '', year: ''};
+    this.handleMonthChange = this.handleMonthChange.bind(this);
+    this.handleDateChange = this.handleDateChange.bind(this);
+    this.handleYearChange = this.handleYearChange.bind(this);
+  }
+
+  componentWillMount() {
+    this.id = _.uniqueId('date-input-');
```

Comment:
Lodash seems to be a common utility library that provides a bunch of standard library features that exist in most sane languages (ruby, python, Java, etc).  Apparently lodash's predecessor was "underscore" (silly punny JS community)

Like jQuery, the convention is to use an terse variable for the library.

`uniqueId` is just a unique-id generator. Each call produces a different string with the prefix. Looking at the impl, it seems to just be updating a counter that is global to the current Javascript script context (so `window`).

`componentWillMount` is called automatically by React.  There are a set of functions that you can override to hook into various lifecycle events for the component.

It seems that they call adding the component to the DOM "mounting" so componentWillMount is the function that is called just before render.  I think there's a componentDidMount which is executed right after the attach.

I bet there's also an unmount call, and that on a navigation away in ReactRouter, the component is actually destroyed...which is why we cannot retrieve state from the component on the final submit because the component's lifecycles will cause them to have been unmounted and destroyed by our navigations.  On the bright side...that means the DOM won't have a bazillion hidden fields.